### ISSUE TITLE
Pluggable thread pool

### ISSUE THREAD

**Issue by bnoordhuis** [libuv/libuv#267]
This is a feature that io.js will need in the near future.  I'm going to self-assign but if someone else wants to work on it, please comment.

Goal: add a hook that lets the libuv user override the built-in thread pool.  I think it'll look something like this:

``` c
struct uv_threadpool_work;  /* Opaque structure. */
typedef int (*uv_threadpool_add_cb)(struct uv_threadpool_work* work);
typedef int (*uv_threadpool_cancel_cb)(struct uv_threadpool_work* work);
/* cancel_cb == NULL means calls to uv_cancel() always fail with UV_EBUSY. */
void uv_threadpool_set_callbacks(uv_threadpool_add_cb add_cb,
                                 uv_threadpool_cancel_cb cancel_cb);
/* User calls this when work is complete. Should probably be safe to call from any thread. */
void uv_threadpool_complete(struct uv_threadpool_work* work);
```

Maybe we want to extend the callbacks with a second argument for passing options, either as flags or, if we want to be extensible, a struct:

``` c
enum uv_threadpool_options_version {
  UV_THREADPOOL_OPTIONS_V1
/* Enum for symbolic debugging, #define for feature detection. */
#define UV_THREADPOOL_OPTIONS_V1 UV_THREADPOOL_OPTIONS_V1
};

struct uv_threadpool_options_v1 {
  int priority;  /* Just an example. */
};

struct uv_threadpool_options {
  enum uv_threadpool_options_version version;
  union {
    struct uv_threadpool_options_v1 v1;
    /* Extend with v2, v3, etc. when the need arises. */
  } u;
};
```

**Comment by saghul** [comment]
> Goal: add a hook that lets the libuv user override the built-in thread pool.

Do you want thos v1.x? Are you still onboard with making something along the lines of https://github.com/libuv/leps/pull/4 for the future++?

**Comment by bnoordhuis** [comment]
I think this feature request covers most of libuv/leps#4:

* configurable thread pool size

* different thread pools for different tasks (not outlined above but could be signaled through options, like the priority)

* uv_queue_work() as a convenience function can stay

**Comment by saghul** [comment]
But it doesn't cover the main point, which is to make the threadpool usage explicit. The user would need to specify on which thread pool to run a given fs operation, for example. And uv_queue_work would morth into taking a threadpool handle argument.

I think both our goals can be achieved if we build an "abstract executor" thing, which the threadpool handle and the custom executor (your proposal) implement. The threadpool handle would be basically an executor with some predefined callbacks, while if you build a custom execurtor, you canset them yourself, just like in your proposal.

Now, my proposal cannot be made for v1.x (breaks API), but yours can (right?) so we can go with it and make the threadpool a handle later on.

WDYT?

**Comment by bnoordhuis** [comment]
> But it doesn't cover the main point, which is to make the threadpool usage explicit.

Right, that's certainly true. :-)

> And uv_queue_work would morth into taking a threadpool handle argument.

morth == morph?

> Now, my proposal cannot be made for v1.x (breaks API), but yours can (right?) so we can go with it and make the threadpool a handle later on.

Correct, and sounds reasonable.

**Comment by saghul** [comment]
> morth == morph?

Yeah.

> Correct, and sounds reasonable.

Kewl, looks like we're on the same page then :-)

**Comment by saper** [comment]
#319 just proposes a simple change to the threadpools to allow more flexibility. I actually just need another pool with maybe configurable number of threads (to avoid exhaustion of threads if using two different families of work queue clients at the same time).

**Comment by btrask** [comment]
In the project I'm working on, I have my own thread pool and run libuv's fs ops in blocking mode (NULL callback). It would be nice to have an option to simply disable libuv's thread pool at compile time, and let the user do their own wrapping.

I'd also like to note that if the thread pool can be guaranteed to only be used from a single main thread, it can be implemented without locking. This guarantee currently isn't supported by libuv because you can have multiple run loops on different threads, but I doubt most projects actually take advantage of that. Last I checked libuv's thread pool uses a condition lock, which (at least theoretically) is slower than necessary and can block the main thread.

**Comment by saghul** [comment]
> In the project I'm working on, I have my own thread pool and run libuv's fs ops in blocking mode (NULL callback). It would be nice to have an option to simply disable libuv's thread pool at compile time, and let the user do their own wrapping.

Do you have any API suggestion? Running blocking fs operations in a thread is actually not safe, because the request is added to the loop active requests queue, which is not thread-safe.

Ben has another PR (#382) where a runtime setting would disable the threadpool, would that work for you?

> I'd also like to note that if the thread pool can be guaranteed to only be used from a single main thread, it can be implemented without locking. This guarantee currently isn't supported by libuv because you can have multiple run loops on different threads, but I doubt most projects actually take advantage of that. 

Any project which uses more than one loop and runs dns or fs operations already takes advantage of it, we cannot make any assumptions...

> Last I checked libuv's thread pool uses a condition lock, which (at least theoretically) is slower than necessary and can block the main thread.

The condition lock is just used to signal workers when work was added to the work queue iff it was empty. In practice I've never seen any issue with this approach, do you have any references I could check?

**Comment by btrask** [comment]
Thanks for the quick response!

> Do you have any API suggestion?

Well, I'm thinking of a compile time flag, like `./configure --disable-threadpool` or something. However, this isn't really that big of a deal to me.

More importantly, last I checked `uv_getaddrinfo` doesn't work in non-blocking mode (with NULL callback). Currently I need to use the libuv thread pool in order to run that one function. I can open a separate issue for this if necessary.

> Running blocking fs operations in a thread is actually not safe, because the request is added to the loop active requests queue, which is not thread-safe.

I create a separate uv run loop for each worker thread. But in blocking mode there is no reason to interact with the run loop at all, is there?

> Any project which uses more than one loop and runs dns or fs operations already takes advantage of it, we cannot make any assumptions...

But a custom run loop can be tailored to the application at hand. And even for applications that use the thread pool from multiple main threads, it would be possible to create multiple thread pools and then avoid further coordination.

Regarding performance: close to a year ago, I wrote a custom VFS (file system wrapper) for SQLite that sat on top of libuv and allowed all of SQLite to run in non-blocking mode. However the performance (without caching) was roughly a factor of 5 slowdown. The overhead for going back and forth to the thread pool for every little read and write made it unworkable. So anything that can be done to speed up the thread pool helps. But for heavy disk workloads, applications either need to use real threads or native AIO (on platforms that support it, only Windows?), not thread pools.

**Comment by saghul** [comment]
> Well, I'm thinking of a compile time flag, like ./configure --disable-threadpool or something. However, this isn't really that big of a deal to me.

I don't like the idea of disabling features at compile time, then we'd need to add a way to know if it has been done, and so on. A runtime setting should do.

> More importantly, last I checked uv_getaddrinfo doesn't work in non-blocking mode (with NULL callback). Currently I need to use the libuv thread pool in order to run that one function. I can open a separate issue for this if necessary.

Not anymore! :-) https://github.com/libuv/libuv/commit/f2bb8d394c06d06ee45e884600466321455751b6

> I create a separate uv run loop for each worker thread. But in blocking mode there is no reason to interact with the run loop at all, is there?

No, there isn't. I *think* we should be able to fix this without much trouble, can you please open an issue?

> But for heavy disk workloads, applications either need to use real threads or native AIO (on platforms that support it, only Windows?), not thread pools.

The story with AIO is not pretty, alas: http://blog.libtorrent.org/2012/10/asynchronous-disk-io/ and things don't seem to have improved much, AFAIK.

> I can't think of a situation where I'd want the thread pool to grow or shrink at runtime. It's better to just set a generous fixed size IMHO. Maybe there is a use case where it would be handy though.

On another PR I suggested that we have both a minimum number of threads and a maximum one, so setting them to the same value should achieve exactly that.

PS: We have a number of open issues / PRs about different threadpool related work, I plan to create a "meta issue" next week aggregating all of them so we (and everyone who wants to) have a better picture.

**Comment by btrask** [comment]
> I don't like the idea of disabling features at compile time, then we'd need to add a way to know if it has been done, and so on. A runtime setting should do.

It can already be disabled at runtime by using `putenv("UV_THREADPOOL_SIZE=0")`, can't it? I don't think anything more complicated is necessary, and there isn't much advantage anyway.

> https://github.com/libuv/libuv/commit/f2bb8d394c06d06ee45e884600466321455751b6

Great news, thanks.

> No, there isn't. I think we should be able to fix this without much trouble, can you please open an issue?

I created an issue, #469.

> The story with AIO is not pretty, alas: http://blog.libtorrent.org/2012/10/asynchronous-disk-io/ and things don't seem to have improved much, AFAIK.

Yeah, I know. I use an approach I call "big workers" where I switch to the thread pool to do I/O and then stay there as long as possible before switching back. This divides the context switch overhead by the number of operations performed. However async network I/O can't be done from the thread pool (since it would be blocked by file I/O) so at least one context switch per "chunk" is necessary.

Non-blocking and blocking mix like oil and water, and unfortunately thread pools don't help as much as one would like.

> On another PR I suggested that we have both a minimum number of threads and a maximum one, so setting them to the same value should achieve exactly that.

I would just set the number of threads to whatever maximum you were going to use, and let the idle threads stay idle. That said I think libuv's current default of 4 threads is too low for many workloads, even on machines with 4 cores or less. It's better to have >1 thread per core so that the kernel scheduler can make better decisions. My application defaults to 16 threads although 8 might be a better default for libuv.

**Comment by CurlyMoo** [comment]
> different thread pools for different tasks

Some further thoughts on this. What about being able to balance the two. So if we have 4 threads and tell the threadpool to use 2 for FS operations and 2 for other work queued by the user and when needed being able to change this balance
```
uv_threadpool_set_balance(2, 2);
uv_threadpool_set_balance(1, 3);
```

Another thing i would love to see is a threadpool that uses a priority queue. This allows me to built a consumer / listener pattern on top of the threadpool in which certain tasks have higer execution priority than others.

An example, a blocking serial device read cannot be set to block for less then 1 millisecond  (`VTIME` set to 1 and `VMIN` set to 0). If we have multiple events waiting to be executed, it would be great the be able to prioritize them so fast work can be executed before the potential slow work. Most of the work triggered by my events are processed in under 100 microsecond so having them executed before the FS work can give the whole program a more responsive feeling.

**Comment by davisjam** [comment]
I'm reasonably familiar with the libuv and Node codebases and I think this is an interesting potential project.

I see that there has been a lot of discussion on this topic.

Some questions:

1. Is anyone working on a PR?
2. Is there an agreed-upon design somewhere?

**Comment by bnoordhuis** [comment]
@davisjam 

1. No.

2. The approach I outlined in the first comment.  It's flexible enough to cover the use cases mentioned in this and other issues, and can be added without breaking backwards compatibility.

**Comment by davisjam** [comment]
@bnoordhuis Thank you.

What's the work process for new libuv featuresl ike this? Should I work on an LEP as a first step?

**Comment by bnoordhuis** [comment]
The LEP-first approach hasn't quite worked out, IMO.  I'd just start with a minimal pull request as a conversation starter.

**Comment by claam** [comment]
Any update about a new LEP or a person working on it ?

**Comment by davisjam** [comment]
@claam See discussion in #1726. I'm not working on it at the moment, currently juggling an internship and another research project.

**Comment by davisjam** [comment]
I am now working on this again.

**Comment by davisjam** [comment]
Since my goal is for this to get into Node...should I target v1.x (ABI compatibility, if this is possible) or v2 (go wild)?

**Comment by cjihrig** [comment]
Your call. We are going to want to get v2 out the door in the coming months so that we don't end up on the hook supporting v1 until Node 12 goes EOL. That said, I wouldn't go too wild if it's going to make upgrading to v2 a huge pain for users.

**Comment by davisjam** [comment]
In #1726 I've proposed what I think is a v1.x-compatible implementation.
